libmscore/rendermidi.cpp

```diff
@@ -731,6 +735,478 @@ bool Score::isSubdivided(ChordRest* chord, int swingUnit)
+    tickspernote = max(tickspernote, mintickspernote);
+
+    // calculate whether to shorten the duration value.
+    if ( tickspernote*( P + B + S) <= maxticks ) {
+        ; // plenty of space to play the notes without changing the requested trill note duration
+    } else if ( tickspernote == mintickspernote ) {
+        return; // the ornament is impossible to implement respecting the minimum duration and all the notes it contains
+    } else {
+        tickspernote = maxticks / (P + B + S) ; // integer division ignoring remainder
+    }
+    int millespernote = space * tickspernote / maxticks ; // rescale duration into per mille
+    
+    // local function:
+    // look ahead in the given vector to see if the current note is the same pitch as the next note or next several notes.
+    // If so, increment the duration by the appropraite note duration, and increment the index, j, to the next note index
+    // of a different pitch.
+    // The total duration of the tied note is returned, and the index is modified.
+    auto tieForward = [millespernote] (int & j, const vector<int> & vec) {
+        int size = vec.size();
+        int duration = millespernote;
+        while ( j < size-1 && vec[j] == vec[j+1] ) {
+            duration += millespernote;
+            j++;
+        }
+        return duration;
+    };
+
+    // local function:
+    auto makeEvent = [note,chord,chromatic,events] (int pitch, int ontime, int duration) {
+        events->append( NoteEvent(chromatic ? pitch : articulationExcursion(note,note,pitch),
+                                  ontime/chord->actualTicks(),
+                                  duration/chord->actualTicks()));
+        return ontime+duration;
+    };
+    // calculate the number of times to repeat the body, and sustain the last note of the body
+    // 1000 = P + numrepeat*B+sustain + S
+    if ( repeatp ) {
+        numrepeat = (space - millespernote*(P + S)) / (millespernote*B);
+    }
+    if ( sustainp ) {
+        sustain   = space - millespernote*(P + numrepeat*B + S);
+    }
+    // render the prefix
+    for (int j=0; j<P; j++) {
+        ontime = makeEvent(prefix[j], ontime, tieForward(j,prefix));
+    }
+    if ( B > 0 ) {
+       // render the body, but not the final repetion
+       for (int r=0; r < numrepeat-1; r++){
+           for (int j=0; j<B; j++) {
+               ontime = makeEvent(body[j], ontime, millespernote);
+           }
+       }
+       // render the final repetion of body, but not the final note of the repition
+       for (int j=0; j<B-1; j++) {
+           ontime = makeEvent(body[j], ontime, millespernote);
+       }
+       // render the final note of the final repeat of body
+       ontime = makeEvent(body[B-1], ontime, millespernote+sustain);
+    }
+    // render the suffix
+    for (int j=0; j<S; j++) {
+        ontime = makeEvent(suffix[j], ontime, tieForward(j,suffix));
+    }
+}
+  
+bool noteHasGlissandop (Note *note) {
```

Comment:
The p is intended to indicate that the function is a boolean predicate function. 
What is the correct way to name a function which answers the question “Does the given note have a glissando?”
I think ? is not a valid character for a function name in C++?


> On 18 May 2015, at 12:52, Nicolas Froment <notifications@github.com> wrote:
> 
> In libmscore/rendermidi.cpp <https://github.com/musescore/MuseScore/pull/1946#discussion_r30497725>:
> 
> > +           }
> > +       }
> > +       // render the final repetion of body, but not the final note of the repition
> > +       for (int j=0; j<B-1; j++) {
> > +           ontime = makeEvent(body[j], ontime, millespernote);
> > +       }
> > +       // render the final note of the final repeat of body
> > +       ontime = makeEvent(body[B-1], ontime, millespernote+sustain);
> > +    }
> > +    // render the suffix
> > +    for (int j=0; j<S; j++) {
> > +        ontime = makeEvent(suffix[j], ontime, tieForward(j,suffix));
> > +    }
> > +}
> > +  
> > +bool noteHasGlissandop (Note *note) {
> what is the 'p' for in noteHasGlissandop ?
> I guess it could be more suited in the Note class?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/1946/files#r30497725>.
>